### ISSUE TITLE
Add distro version check to scripts in script-library

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-dev-containers#1019]
Relates to: Both

Currently scripts in the `script-library` are setup to assume that support exists for all the major LTS versions of Debian and Ubuntu. However, we've seen a few lag behind (notably the Azure CLI which doesn't yet have a `bullseye` package).

As a result we should add a distro check in scripts that would be affected. As we look at what we've been calling the "lego block" experience or "container features" (https://github.com/microsoft/vscode-remote-release/issues/5442), we can expect to see more mismatching than we have now and should provide good error messages when it happens.

**Comment by joshspicer** [comment]
debating how I should do this. 

1. `apt-cache policy azure-cli | less` and try to parse out the `$VERSION_CODENAME` I get from /etc/os-release
2. Keep a "fail-list" that we maintain of LTS distros that we know break the script (eg: `azcli` fails if run when `$VERSION_CODENAME` == `bullseye`)
3. Some other, better/more obvious way

Did you have an idea how to accomplish this @chuxel?

**Comment by Chuxel** [comment]
Yeah it probably varies a bit depending on how the bits are acquired.

> 1. `apt-cache policy azure-cli | less` and try to parse out the `$VERSION_CODENAME` I get from /etc/os-release
> 2. Keep a "fail-list" that we maintain of LTS distros that we know break the script (eg: `azcli` fails if run when `$VERSION_CODENAME` == `bullseye`)
> 3. Some other, better/more obvious way

The thing I notice is that every major release seems to require subtle updates to scripts that invariably catches me by suprise.  Bullseye for example doesn't have a `python` command by default when you install `python3-minimal` unless you install `python-is-python3`.  That caused problems with `node-gyp` in the Node definition much to my surprise.  Different versions of the script therefore will have different sets of distros they support.

I suspect we need to more tightly control when we switch to new distro versions and run a script regression on them. The arm64 changes I just made do that for the images themselves (we control what our MCR python:3 image is). We likely need to have an "allow" list rather than a "it doesn't work" list to be safe.

This will probably be even more important as we add "features" into the mix since people can mix and match any combination of things.

**Comment by joshspicer** [comment]
tying this into the devcontainerv2 work, this could be useful as a more generic feature script function as a "sanity check"

**Comment by joshspicer** [comment]
>  \> (notably the Azure CLI which doesn't yet have a bullseye package).

I have no problem now running the `./azcli-debian.sh` script in a bullseye container.

```bash
➜ /workspaces/vscode-dev-containers/script-library (dev/jospicer/update-scripts-0930) $ docker run --rm -it --entrypoint=/bin/bash -v `pwd`:/here -w /here mcr.microsoft.com/vscode/devcontainers/base:bullseye

Unable to find image 'mcr.microsoft.com/vscode/devcontainers/base:bullseye' locally
bullseye: Pulling from vscode/devcontainers/base
4c25b3090c26: Pull complete 
1acf565088aa: Pull complete 
b95c0dd0dc0d: Pull complete 
bb05c398806d: Pull complete 
a62153dbcdcf: Pull complete 
Digest: sha256:d7240de2fe2b8a1016e0c92a9221dd6e3121180ee84feee54e0e9a4e5a6aa7a8
Status: Downloaded newer image for mcr.microsoft.com/vscode/devcontainers/base:bullseye

root ➜ /here $ ./azcli-debian.sh 

(*) Installing Azure CLI...
MICROSOFT_GPG_KEYS_URI=https://packages.microsoft.com/keys/microsoft.asc
Get:1 http://deb.debian.org/debian bullseye InRelease [113 kB]
Get:2 http://deb.debian.org/debian bullseye-updates InRelease [39.4 kB]
Get:3 https://packages.microsoft.com/repos/azure-cli bullseye InRelease [10.4 kB]
Get:4 http://security.debian.org/debian-security bullseye-security InRelease [44.1 kB]
Get:5 http://deb.debian.org/debian bullseye/non-free amd64 Packages [93.8 kB]
Get:6 http://deb.debian.org/debian bullseye/contrib amd64 Packages [50.4 kB]
Get:7 http://deb.debian.org/debian bullseye/main amd64 Packages [8178 kB]
Get:8 http://deb.debian.org/debian bullseye-updates/main amd64 Packages [2300 B]
Get:9 https://packages.microsoft.com/repos/azure-cli bullseye/main amd64 Packages [613 B]
Get:10 http://security.debian.org/debian-security bullseye-security/main amd64 Packages [41.0 kB]
Fetched 8572 kB in 1s (7230 kB/s)                           
Reading package lists... Done
Reading package lists...
Building dependency tree...
Reading state information...
The following NEW packages will be installed:
  azure-cli
0 upgraded, 1 newly installed, 0 to remove and 1 not upgraded.
Need to get 64.4 MB of archives.
After this operation, 955 MB of additional disk space will be used.
Get:1 https://packages.microsoft.com/repos/azure-cli bullseye/main amd64 azure-cli all 2.28.0-1~bullseye [64.4 MB]
Fetched 64.4 MB in 1s (95.3 MB/s)
Selecting previously unselected package azure-cli.
(Reading database ... 18956 files and directories currently installed.)
Preparing to unpack .../azure-cli_2.28.0-1~bullseye_all.deb ...
Unpacking azure-cli (2.28.0-1~bullseye) ...
Setting up azure-cli (2.28.0-1~bullseye) ...
Done!
```

**Comment by Chuxel** [comment]
> tying this into the devcontainerv2 work, this could be useful as a more generic feature script function as a "sanity check"

Yes, sounds good.

> > (notably the Azure CLI which doesn't yet have a bullseye package).
>
> I have no problem now running the ./azcli-debian.sh script in a bullseye container.

That's because I raised an issue to get one created. :smile: 

However, even here, there wasn't an arm64 package, so that's why I added the pip install falback (based on the response to my raised issue).

When bullseye first released, the script just bombed.

**Comment by joshspicer** [comment]
Yep, agree that we definitely need some kind of check.  I want to hopefully find something that doesn't take a ton of work to maintain

**Comment by Chuxel** [comment]
@joshspicer Yeah, one of the reasons I added that "settings.env" file was to allow us to have a spot to dynamically update these kinds of lists. The script contains a default set but would be overridden by a central file.  We can probably harden where we put it to avoid typos causing problems, but the general idea was to help with some of this.  It's already getting used to dictate when fallbacks and errors happen in a few spots.
